File: Godot-WRO-Future-Engineers-2024-Simulator/PythonScripts/pidcontroller.py
class PIDController:
    def __init__(self, kp, ki, kd):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.previous_error = 0
        self.integral = 0

    def update(self, error, delta_time):
        # Proportional term
        p = self.kp * error
        
        # Integral term
        self.integral += error * delta_time
        i = self.ki * self.integral
        
        # Derivative term
        if delta_time > 0:
            derivative = (error - self.previous_error) / delta_time
        else:
            derivative = 0
        d = self.kd * derivative
        
        # Save current error for next derivative calculation
        self.previous_error = error
        
        # Return the PID output
        return p + i + d

File: RaspberryPi5/RaspberryPi5-Code/src/utils/PIDController.cpp
class PIDController {
public:
    PIDController(float kp, float ki, float kd) 
        : kp(kp), ki(ki), kd(kd), prevError(0.0f), integral(0.0f) {}

    float calculate(float error, float deltaTime) {
        integral += error * deltaTime;
        float derivative = (error - prevError) / deltaTime;
        prevError = error;

        return kp * error + ki * integral + kd * derivative;
    }

private:
    float kp;         // Proportional coefficient
    float ki;         // Integral coefficient
    float kd;         // Derivative coefficient
    float prevError;  // Previous error
    float integral;   // Accumulated integral
};
